doc/designs/ddd/Makefile

```diff
@@ -1,26 +1,36 @@
 #
-# To run the demos when linked with a shared library (default):
+# To run the demos when linked with a shared library (default) ensure that
+# libcrypto and libssl are on the library path. For example to run the
+# ddd-01-conn-blocking-tls demo:
 #
-#    LD_LIBRARY_PATH=../.. make test
+#    LD_LIBRARY_PATH=../../.. ./ddd-01-conn-blocking-tls
+#
+# Building ddd-06-mem-uv-tls and ddd-06-mem-uv-quic requires the
+# library libuv and header file.  On Ubuntu, they are provided by the
+# package "libuv1-dev".
+
+TESTS_BASE = ddd-01-conn-blocking \
+             ddd-02-conn-nonblocking \
+             ddd-02-conn-nonblocking-threads \
+             ddd-03-fd-blocking \
+             ddd-04-fd-nonblocking \
+             ddd-05-mem-nonblocking \
+             ddd-06-mem-uv
 
-TESTS_BASE=ddd-01-conn-blocking ddd-02-conn-nonblocking ddd-02-conn-nonblocking-threads \
-		   ddd-03-fd-blocking ddd-04-fd-nonblocking ddd-05-mem-nonblocking ddd-06-mem-uv
-TESTS=$(foreach x,$(TESTS_BASE),$(x)-tls $(x)-quic)
+TESTS = $(foreach x,$(TESTS_BASE),$(x)-tls $(x)-quic)
 
-CFLAGS = -I../../../include -O3 -g -Wall
-LDFLAGS = -L../../..
-LDLIBS = -lcrypto -lssl
+CFLAGS      = -I../../../include -g -Wall -Wsign-compare
+LDFLAGS     = -L../../..
+LDLIBS_BASE = -lcrypto -lssl
+LDLIBS      = $(LDLIBS_BASE)
 
 all: $(TESTS)
 
 clean:
 	rm -f $(TESTS) *.o
 
-ddd-06-mem-uv-tls: ddd-06-mem-uv.c
-	$(CC) $(CFLAGS) $(LDFLAGS) -o "$@" "$<" $(LDLIBS) -luv
-
-ddd-06-mem-uv-quic: ddd-06-mem-uv.c
-	$(CC) $(CFLAGS) $(LDFLAGS) -DUSE_QUIC -o "$@" "$<" $(LDLIBS) -luv
+ddd-06-mem-uv-quic: LDLIBS = $(LDLIBS_BASE) -luv
+ddd-06-mem-uv-tls:  LDLIBS = $(LDLIBS_BASE) -luv
```

Comment:
These two don't look right.  Are they a GNUMake-ism?

Reply:
In GNU make, this is called "target specific variable values":

https://www.gnu.org/software/make/manual/make.html#Target_002dspecific

I am not certain how well it is supported by other versions of make.  It may very well be a GNU make-ism.

This makefile uses pattern rules (i.e. rules that use `%`), so it was already broken for make versions that do not have that (like nmake).

The thing I wanted to make more clear in the makefile was that all the demos are built the same way, except that demo `ddd-06` needs to link against `libuv`.

We could just use the same `LDLIBS` options for building each ddd demo, and that would simpler.

Reply:
We should avoid GNU specific stuff IMO.

Pattern rules date way way back, I'm less concerned about them.  Other Makefiles in the demos area use them, so redressing this should be a separate PR IMO.

Reply:
understood.

The question remains whether or not target specific variable values is GNU make specific.  I do not have enough experience with other make implementations to judge that.

I found that Ubuntu has a NetBSD package.  I installed it, but I cannot get "bmake" to build anything.  It does not seem to support pattern rules.

Reply:
I've just pushed an updated.  Let me know if you are okay with this version of the makefile.